x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,405 @@
+			s3Info := s3Info{}
+			s3Info.region = recordMap["awsRegion"].(string)
+			s3Record := recordMap["s3"].(map[string]interface{})
+
+			bucketInfo := s3Record["bucket"].(map[string]interface{})
+			s3Info.name = bucketInfo["name"].(string)
+
+			objectInfo := s3Record["object"].(map[string]interface{})
+			s3Info.key = objectInfo["key"].(string)
+			s3Infos = append(s3Infos, s3Info)
+		}
+	}
+	return s3Infos, nil
+}
+
+// stringInSlice checks if a string is already exists in list
+// If there is no bucketNames configured, then collect all.
+func stringInSlice(name string, bucketNames []string) bool {
+	if bucketNames == nil || len(bucketNames) == 0 {
+		return true
+	}
+
+	for _, v := range bucketNames {
+		if v == name {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *Input) readS3CreateEvents(svc s3iface.S3API, s3Infos []s3Info) {
+	if len(s3Infos) > 0 {
+		var wg sync.WaitGroup
+		numS3Infos := len(s3Infos)
+		wg.Add(numS3Infos)
+
+		for i := range s3Infos {
+			go func(s3Info s3Info) {
+				// launch goroutine to handle each message
+				defer wg.Done()
+
+				// read from s3 object
+				reader, err := bufferedIORead(svc, s3Info)
+				if err != nil {
+					p.logger.Error(errors.Wrap(err, "s3 get object request failed"))
+					return
+				}
+
+				offset := 0
+				for {
+					log, err := reader.ReadString('\n')
+					if log == "" {
+						break
+					}
+
+					if err != nil {
+						if err == io.EOF {
+							offset += len([]byte(log))
+							p.forwardEvent(createEvent(log, offset, s3Info))
+							break
+						} else {
+							p.logger.Error(errors.Wrap(err, "ReadString failed"))
+						}
+					}
+
+					// create event per log line
+					offset += len([]byte(log))
+					p.forwardEvent(createEvent(log, offset, s3Info))
+				}
+			}(s3Infos[i])
+		}
+		wg.Wait()
+	}
+}
+
+func bufferedIORead(svc s3iface.S3API, s3Info s3Info) (*bufio.Reader, error) {
+	s3GetObjectInput := &s3.GetObjectInput{
+		Bucket: awssdk.String(s3Info.name),
+		Key:    awssdk.String(s3Info.key),
+	}
+	req := svc.GetObjectRequest(s3GetObjectInput)
+
+	resp, err := req.Send()
+	if err != nil {
+		return nil, errors.Wrap(err, "s3 get object request failed")
+	}
+
+	return bufio.NewReader(resp.Body), nil
+}
+
+func (p *Input) forwardEvent(event *beat.Event) {
+	forwarder := harvester.NewForwarder(p.outlet)
+	d := &util.Data{Event: *event}
+	err := forwarder.Send(d)
+	if err != nil {
+		p.logger.Error(errors.Wrap(err, "forwarder send failed"))
+	}
+}
+
+func deleteMessage(queueURL string, messagesReceiptHandle string, svcSQS *sqs.SQS) error {
+	deleteMessageInput := &sqs.DeleteMessageInput{
+		QueueUrl:      awssdk.String(queueURL),
+		ReceiptHandle: awssdk.String(messagesReceiptHandle),
+	}
+
+	req := svcSQS.DeleteMessageRequest(deleteMessageInput)
+	_, err := req.Send()
+	if err != nil {
+		return errors.Wrap(err, "DeleteMessageRequest failed")
+	}
+	return nil
+}
+
+func createEvent(log string, offset int, s3Info s3Info) *beat.Event {
+	f := common.MapStr{
```

Comment:
Given that key and version of S3 objects uniquely identify them within a bucket you could assign each event a unique ID to prevent duplicate events. The key could be `sha256(bucket + key + version)[:10] + "-" offset` (if you follow my shorthand) (e.g. 8a9f49af1b-6137). You would set `beat.Event.Meta["id]` with that value.

This would prevent duplicates later if Filebeat is interrupted while processing an object and does not `DeleteMessage`.